### DISCUSSION TITLE
request for exposing joint current angle / linear distance

### DISCUSSION THREAD

**Discussion by LunarShadow01** [godot-jolt/godot-jolt#598] [General]
would it be much to ask that the angle / linear distance for the joints would be exposed

I currently see no way to have accurately stop a joint at a specific angle / distance
the only option i can think about is to motorize it for angle / velocity seconds
and after the time is up set the velocity to zero to so the motor would hold the position
but without setting a the limits to the current angle the motor might slip

and this is not as accurate as reading the angle / distance from the engine and then setting limits
or motorizing it in that direction for correction

**Comment by mihe** [comment]
I'm not sure I understand your use-case.

If you know the angle/distance at which you want to limit the joint, why wouldn't you just set the actual limit of the joint to that value? Or do you not in fact know the angle/distance and you just want the joint to stop at some particular point regardless of what the angle/distance is?

**Reply by LunarShadow01** [reply]
the problem with simply setting the angle/distance in the limit parameter is that it will instantly snap to it
and if i don't know the current angle of the joint i also can't increase the limit in the appropriate upper/lower limits as i don't know if my new angle is higher or lower than the current angle

(for clarification. if i were to know my current angle and the angle i want to reach i can simply set the bounds to include my current and desired angle. the motorize it until the joint reaches that upper limit. and i would have no change of over shooting)

my use case is to have the joint go from whichever point it is at and motorize it to the new point
this can probably be achieved with timing. by taking the desired angle i want to add divided by the velocity i will set my joint to
and that would give me the time the joint needs to move at that velocity to reach the new angle

the problem with that approach is i don't know how accurate it would be and it doesn't work if i want to modify the speed half way through the process

i don't know how complex it would be to expose the angle/distance but i believe it would be useful for doing more complex sequences of movement via code. more accurately. and provide more use cases for the joints

**Reply by mihe** [reply]
Which set of joints are you using? The default ones (i.e. `HingeJoint3D` etc.) or the substitute ones (i.e. `JoltHingeJoint3D`. etc.)?

> i don't know how complex it would be to expose the angle/distance but i believe it would be useful for doing more complex sequences of movement via code. more accurately. and provide more use cases for the joints

It's not something I can do for the default set of joints, since that interface is defined on the Godot side of things. I can however add it to the substitute joints, and it is indeed something that Jolt exposes on its constraints/joints.

The problem is that [I mess around](https://github.com/godot-jolt/godot-jolt/blob/b6f68dc32c935826526dda6e87cc61ad4df74f7e/src/joints/jolt_hinge_joint_impl_3d.cpp#L280-L298) with the bodies relative transforms in order to accommodate for Godot's ability to set asymmetrical limits, meaning limits like [-15, 25] or even [-15, -15], as opposed to symmetrical limits like [-15, 15], which I believe will skew whatever current angle/distance that Jolt reports to me.

I'll have to think about this for a bit.

> my use case is to have the joint go from whichever point it is at and motorize it to the new point

If you don't necessarily need the fixed speed when you go between the two points, you can maybe also achieve this using a 6DOF spring. If you increase the iterations under "Solver Overrides" to something like 10+ and crank up the spring frequency fairly high, with maybe some damping as well, then you can use the "Equilibrium Point" property on each axis to steer where the bodies should go. I'm not sure if it's accurate enough for you, but might be worth giving a shot.

**Reply by LunarShadow01** [reply]
> It's not something I can do for the default set of joints, since that interface is defined on the Godot side of things. I can however add it to the substitute joints, and it is indeed something that Jolt exposes on its constraints/joints.

I am using the jolt joints. they are more consistent for my project's needs

> I'll have to think about this for a bit.

no problem. I'm in no rush

> If you don't necessarily need the fixed speed when you go between the two points, you can maybe also achieve this using a 6DOF spring. If you increase the iterations under "Solver Overrides" to something like 10+ and crank up the spring frequency fairly high, with maybe some damping as well, then you can use the "Equilibrium Point" property on each axis to steer where the bodies should go. I'm not sure if it's accurate enough for you, but might be worth giving a shot.

fixed speed would be nice but its not a strict requirement at the moment
the idea of using 6DOF joint is interesting. i'm already using it so i can create any custom joint I would want
I'll try to look into it more tommorow as I'm unfamiliar with the the spring functionality and its various parameters
however the solver override part does sound like it would impact performance
If my understanding is correct it would add more steps to the calculations that are done each frame

regardless thanks for the help. whether or not the parameters would be exposed this is some good thinking material

**Reply by mihe** [reply]
> however the solver override part does sound like it would impact performance

Yes, it will, but in a more localized way than setting it globally in the project settings. It might not even be necessary though, but from a quick test it seemed like the joint would tend to overshoot or jitter as it reached the target if it wasn't given more iterations. Maybe some damping could have alleviated that though.

**Reply by LunarShadow01** [reply]
I have just started doing a bit of testing.
I have not tried your suggestion of using higher solver steps with a spring damping etc
but I have confirmed something which came up to my mind yesterday

the system of angle divided by speed works even without the extra steps
but it doesn't work when you introduce obstructions or friction to the mix

both of these would slow down or completely stop the joints from rotating but the timer will still run
the result of this is such that the joint will not travel the desired angle if it is slowed down by anything at any point
for any amount of time.

having access to the joints current angle / distance would solve this as i'll be able to know how much i've traveled and correct for it with more logic

apart from that, my test scenario suggests that for in perfect conditions the angle divided by speed method does work quite accurately

**Comment by i-draw-boxes** [comment]
Since you refered me here, I might add what my usecase is for having the same request:

I have two rigid bodies joined by a 6DOF joint that is controlled (as you suggest in the previous responses to this) by the angular spring settings. This lets me position it exactly as I want. 

However, I am not able to "reset" the springs to a new resting point if the arm is moved by external forces. Therefore it would be nice to get the current offset or absolute angle for each joint axis so I can set the resting point of the springs to that angle, effectively removing all tension from the joint and stabilizing it in the current configuration.

**Reply by LunarShadow01** [reply]
Yes having access the off or absolute angle simply eliminates a lot of variability you otherwise cannot account for. Which is why i would like for it to be exposed

**Reply by i-draw-boxes** [reply]
I think the list of new properties could include the following:

- Current applied linear impulse as a vector (only the magnitude of it is available as a single float)
- Current applied angular impulse as quaternion or euler angles (only the magnitude of it is available as a single float, not sure if there are issues with euler angles in this context)
- Offset and absolute angles for each individual joint axis
- Offset and absolute distances for each individual joint axis
- Current linear motor speeds for each axis
- Current angular motor speeds for each axis
- Absolute position of the joint in world space as seen by the physics engine (I think one can currently approximate this by taking the initial offset to body1 in body1 local space and transforming it into world space later)

I know some of these properties are very similar or can be derived from each other with simple calculations, nonetheless it would be nice to have them directly available.

**Reply by mihe** [reply]
> * Current applied linear impulse as a vector (only the magnitude of it is available as a single float)
> * Current applied angular impulse as quaternion or euler angles (only the magnitude of it is available as a single float, not sure if there are issues with euler angles in this context)

These were deliberately made into a magnitude due to how Jolt reports these things, as discussed [here](https://github.com/jrouwe/JoltPhysics/discussions/629).

> * Offset and absolute angles for each individual joint axis
> * Offset and absolute distances for each individual joint axis

As touched upon [already](https://github.com/godot-jolt/godot-jolt/discussions/598#discussioncomment-7061970), while Jolt offers this as part of its API, I don't believe this is all that straight-forward with the way I've hacked in support for asymmetrical limits.

> * Current linear motor speeds for each axis
> * Current angular motor speeds for each axis

I don't believe this is exposed by Jolt, if it even makes sense in the first place. The only thing I see being available is the impulse needed to satisfy the motor constraint part, which I presume would suffer from the same problem as the offsets mentioned above, meaning making more sense as a magnitude rather than a per-axis value.

> Absolute position of the joint in world space as seen by the physics engine (I think one can currently approximate this by taking the initial offset to body1 in body1 local space and transforming it into world space later)

The initial transforms of the two bodies relative to the joint are [readily available](https://github.com/godot-jolt/godot-jolt/blob/b0866be9b6f1d6114ce75cf7440a44960516ab10/src/joints/jolt_joint_impl_3d.hpp#L88-L90), so this should be fairly trivial, and seems like something that could be useful... I think.

**Comment by GeorgeS2019** [comment]
Could someone add either c#  in the title or add a topic or label for this issue using label : Csharp

Jolt could be the future and I hope c# is taken care, especially for unity migrants

FYI: https://github.com/godot-jolt/godot-jolt/issues/638

## Other c# issues
https://github.com/godot-jolt/godot-jolt/discussions/243

## Examples of Godot Jolt for c#

- https://github.com/jamesdimick/godot-jolt-issue-repro
- https://github.com/KuroiRoy/GodotJoltCrashMRP
- https://github.com/jamesdimick/godot-jolt-issue-repro

**Reply by mihe** [reply]
Just to be clear, this is not about making this information more readily available from C# specifically. If anything this would be more tedious to access from C# compared to GDScript, since you'd have to access it all through `GodotObject.Get` and `GodotObject.Call`, due to there being no C# bindings for this extension (#632).

**Comment by tlobig** [comment]
Hi @mihe I'd like to unearth this discussion again. I might just be dense but from how I  understand your responses the current angle on a joint axis is already exposed through the API. I can't figure out how. So if it's really already available, can you be specific about how to access this value, please?

Just for context: I have pairs of joints that do a revolving motion and I want to synch them up. It's actually some kind of robotic creature with legs and I use the angular motor to turn them. It looks fun, but they are getting out of synch soon with just the ability to set a target velocity, but no feedback where they are at.
I would need to have access to the same angle the "motor" is using to control the angular velocity versus the target velocity.
I would either speed up or slow down this target speed based on how much the pairwise angles are out of synch. A basic PID.

**Reply by tlobig** [reply]
For the time being I also do this with angular spring instead of motor. I would rather want to set a max torque instead of frequency and damping, but alas at least in my case I do at least have mostly the behaviour I need.
